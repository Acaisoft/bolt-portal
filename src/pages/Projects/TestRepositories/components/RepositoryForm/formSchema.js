import * as Yup from 'yup'

import { makeValidationSchema, makeEmptyInitialValues } from '~utils/forms'

const formFields = {
  name: {
    label: 'Repository Name',
    validator: Yup.string()
      .required()
      .ensure()
      .min(2),
  },
  url: {
    label: 'Repository URL',
    validator: Yup.string()
      .required()
      .ensure()
      .min(2),
  },
}

const validationSchema = makeValidationSchema(formFields)
const initialValues = makeEmptyInitialValues(formFields)

export { formFields, validationSchema, initialValues }
