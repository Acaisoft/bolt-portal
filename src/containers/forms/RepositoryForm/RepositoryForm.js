import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Formik } from 'formik'
import { Mutation } from 'react-apollo'

import { toast } from 'react-toastify'
import { Button, withStyles } from '@material-ui/core'
import { FormField } from '~components'

import { formFields, validationSchema } from './formSchema'
import styles from './RepositoryForm.styles'

import {
  ADD_REPOSITORY_MUTATION,
  EDIT_REPOSITORY_MUTATION,
} from '~services/GraphQL/Mutations'
import { GET_REPOSITORIES_QUERY } from '~services/GraphQL/Queries'

export class RepositoryForm extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    initialValues: PropTypes.object,
    mode: PropTypes.oneOf(['create', 'edit']),
    onSubmit: PropTypes.func,
    projectId: PropTypes.string,
  }

  static defaultProps = {
    mode: 'create',
    onSubmit: () => {},
  }

  handleSubmit = (values, { repoMutation }) => {
    const { id, name, url } = values
    try {
      const variables = {
        name,
        url,
      }

      if (this.props.mode === 'create') {
        variables.projectId = this.props.projectId
      } else {
        variables.id = id
      }

      repoMutation({ variables })
    } catch (err) {
      toast.error('An error has occured. Project was not created.')
    }

    this.props.onSubmit(values)
  }

  render() {
    const { initialValues, onCancel, mode } = this.props
    return (
      <Mutation
        mutation={
          mode === 'create' ? ADD_REPOSITORY_MUTATION : EDIT_REPOSITORY_MUTATION
        }
        refetchQueries={[{ query: GET_REPOSITORIES_QUERY }]}
      >
        {(repoMutation, { data }) => (
          <Formik
            initialValues={initialValues}
            onSubmit={values => this.handleSubmit(values, { repoMutation })}
            validationSchema={validationSchema}
          >
            {({ handleSubmit, isSubmitting, dirty }) => (
              <form onSubmit={handleSubmit}>
                {Object.entries(formFields).map(([name, options]) => {
                  return (
                    <FormField
                      key={name}
                      name={name}
                      label={options.label}
                      margin="normal"
                      fullWidth
                      {...options.input}
                    />
                  )
                })}
                <Button
                  color="primary"
                  variant="contained"
                  onClick={this.handleModalOpen}
                >
                  GENERATE KEY
                </Button>

                <Button
                  color="primary"
                  variant="text"
                  disabled={isSubmitting}
                  onClick={() => onCancel()}
                >
                  Cancel
                </Button>
                <Button
                  color="primary"
                  variant="contained"
                  type="submit"
                  disabled={!dirty || isSubmitting}
                >
                  {mode === 'create' ? 'Add repository' : 'Update'}
                </Button>
              </form>
            )}
          </Formik>
        )}
      </Mutation>
    )
  }
}

export default withStyles(styles)(RepositoryForm)
