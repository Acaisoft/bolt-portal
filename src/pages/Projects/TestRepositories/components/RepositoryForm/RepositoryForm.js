import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Formik } from 'formik'
import { Mutation } from 'react-apollo'
import { toast } from 'react-toastify'

import { withStyles } from '@material-ui/core/styles'

import Drawer from '@material-ui/core/Drawer'
import Button from '@material-ui/core/Button'
import AppBar from '@material-ui/core/AppBar'
import Typography from '@material-ui/core/Typography'

import FormField from '~components/FormField'
import { formFields, validationSchema } from './formSchema'

import styles from './RepositoryForm.styles'

import { ADD_REPOSITORY_MUTATION } from '~services/GraphQL/Mutations'
import { EDIT_REPOSITORY_MUTATION } from '~services/GraphQL/Mutations'
import { GET_REPOSITORIES_QUERY } from '~services/GraphQL/Queries'

export class RepositoryForm extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    open: PropTypes.bool.isRequired,
    close: PropTypes.func.isRequired,
    courseInitData: PropTypes.object,
  }

  handleSubmit = (values, { repoMutation }) => {
    const { name, url } = values
    try {
      if (this.props.type === 'create') {
        repoMutation({ variables: { name, url } }) //TODO: Add projectId
      } else {
        repoMutation({
          variables: { name, url, id: values.id },
        })
      }
    } catch (err) {
      toast.error('An error has occured. Project was not created.')
    }
    this.props.close(null, false)
  }

  renderForm = ({ handleSubmit, isSubmitting, dirty }) => (
    <React.Fragment>
      <AppBar className={this.props.classes.appBar} position="static">
        <Typography variant="h3">
          {this.props.type === 'create' ? 'NEW REPOSITORY' : 'UPDATE REPOSITORY'}
        </Typography>
        <Typography variant="body1">
          Here you can define your new test run repository, you can edit the data
          until first test run. After that you will be able only to change
          configuration name.
        </Typography>
      </AppBar>
      <form onSubmit={handleSubmit}>
        {Object.entries(formFields).map(([name, options]) => {
          return (
            <FormField
              key={name}
              name={name}
              label={options.label}
              margin="normal"
              fullWidth
              {...options.input}
            />
          )
        })}

        <Button
          color="primary"
          variant="text"
          disabled={isSubmitting}
          onClick={() => this.props.close(null, false)}
        >
          Cancel
        </Button>
        <Button
          color="primary"
          variant="contained"
          type="submit"
          disabled={!dirty || isSubmitting}
        >
          {this.props.type === 'create' ? 'ADD REPOSITORY' : 'UPDATE'}
        </Button>
      </form>
    </React.Fragment>
  )

  render() {
    const { classes, open, courseInitData, type } = this.props
    return (
      <div>
        <Drawer
          open={open}
          anchor="right"
          classes={{
            paper: classes.drawer,
          }}
        >
          <Mutation
            mutation={
              type === 'create' ? ADD_REPOSITORY_MUTATION : EDIT_REPOSITORY_MUTATION
            }
            refetchQueries={[{ query: GET_REPOSITORIES_QUERY }]}
          >
            {(repoMutation, { data }) => (
              <Formik
                initialValues={courseInitData}
                onSubmit={values => this.handleSubmit(values, { repoMutation })}
                validationSchema={validationSchema}
              >
                {this.renderForm}
              </Formik>
            )}
          </Mutation>
        </Drawer>
      </div>
    )
  }
}

export default withStyles(styles)(RepositoryForm)
