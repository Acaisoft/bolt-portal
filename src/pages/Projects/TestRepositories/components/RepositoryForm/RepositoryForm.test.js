import React from 'react'
import { shallow } from 'enzyme'

import { RepositoryForm } from './RepositoryForm'

import { mocks } from '~utils/tests'

const initComponent = overrides => {
  const mockProps = {
    open: false,
    close: jest.fn(),
  }
  const wrapper = shallow(
    <RepositoryForm {...mockProps} {...overrides} classes={mocks.ClassesProxy} />
  )
  const instance = wrapper.instance()
  return { instance, mockProps, wrapper }
}

describe('component Projects/List - RepositoryForm', () => {
  describe('rendering', () => {
    it('should render without crashing', () => {
      const { wrapper } = initComponent()
      expect(wrapper).toBeTruthy()
    })
  })
  describe('events', () => {
    describe('handleSubmit', () => {
      it('should close drawer', () => {
        const { instance, mockProps } = initComponent()
        const values = {}
        const addProjectMock = jest.fn()
        instance.handleSubmit(values, { addProjectMock })
        expect(mockProps.close).toHaveBeenCalled()
      })
    })
  })
})
