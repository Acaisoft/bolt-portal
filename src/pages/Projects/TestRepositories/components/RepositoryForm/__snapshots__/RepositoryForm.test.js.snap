// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: RepositoryForm rendering should render as expected 1`] = `
<div>
  <KeyRepositoryModal
    handleClose={[Function]}
    open={false}
  />
  <WithStyles(Drawer)
    anchor="right"
    classes={
      Object {
        "paper": "drawer",
      }
    }
    open={false}
  >
    <Mutation
      mutation={
        Object {
          "definitions": Array [
            Object {
              "directives": Array [],
              "kind": "OperationDefinition",
              "name": undefined,
              "operation": "mutation",
              "selectionSet": Object {
                "kind": "SelectionSet",
                "selections": Array [
                  Object {
                    "alias": Object {
                      "kind": "Name",
                      "value": "repository",
                    },
                    "arguments": Array [
                      Object {
                        "kind": "Argument",
                        "name": Object {
                          "kind": "Name",
                          "value": "_set",
                        },
                        "value": Object {
                          "fields": Array [
                            Object {
                              "kind": "ObjectField",
                              "name": Object {
                                "kind": "Name",
                                "value": "name",
                              },
                              "value": Object {
                                "kind": "Variable",
                                "name": Object {
                                  "kind": "Name",
                                  "value": "name",
                                },
                              },
                            },
                            Object {
                              "kind": "ObjectField",
                              "name": Object {
                                "kind": "Name",
                                "value": "url",
                              },
                              "value": Object {
                                "kind": "Variable",
                                "name": Object {
                                  "kind": "Name",
                                  "value": "url",
                                },
                              },
                            },
                          ],
                          "kind": "ObjectValue",
                        },
                      },
                      Object {
                        "kind": "Argument",
                        "name": Object {
                          "kind": "Name",
                          "value": "where",
                        },
                        "value": Object {
                          "fields": Array [
                            Object {
                              "kind": "ObjectField",
                              "name": Object {
                                "kind": "Name",
                                "value": "id",
                              },
                              "value": Object {
                                "fields": Array [
                                  Object {
                                    "kind": "ObjectField",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "_eq",
                                    },
                                    "value": Object {
                                      "kind": "Variable",
                                      "name": Object {
                                        "kind": "Name",
                                        "value": "id",
                                      },
                                    },
                                  },
                                ],
                                "kind": "ObjectValue",
                              },
                            },
                          ],
                          "kind": "ObjectValue",
                        },
                      },
                    ],
                    "directives": Array [],
                    "kind": "Field",
                    "name": Object {
                      "kind": "Name",
                      "value": "update_repository",
                    },
                    "selectionSet": Object {
                      "kind": "SelectionSet",
                      "selections": Array [
                        Object {
                          "alias": undefined,
                          "arguments": Array [],
                          "directives": Array [],
                          "kind": "Field",
                          "name": Object {
                            "kind": "Name",
                            "value": "returning",
                          },
                          "selectionSet": Object {
                            "kind": "SelectionSet",
                            "selections": Array [
                              Object {
                                "alias": undefined,
                                "arguments": Array [],
                                "directives": Array [],
                                "kind": "Field",
                                "name": Object {
                                  "kind": "Name",
                                  "value": "id",
                                },
                                "selectionSet": undefined,
                              },
                              Object {
                                "alias": undefined,
                                "arguments": Array [],
                                "directives": Array [],
                                "kind": "Field",
                                "name": Object {
                                  "kind": "Name",
                                  "value": "name",
                                },
                                "selectionSet": undefined,
                              },
                              Object {
                                "alias": undefined,
                                "arguments": Array [],
                                "directives": Array [],
                                "kind": "Field",
                                "name": Object {
                                  "kind": "Name",
                                  "value": "url",
                                },
                                "selectionSet": undefined,
                              },
                            ],
                          },
                        },
                      ],
                    },
                  },
                ],
              },
              "variableDefinitions": Array [
                Object {
                  "defaultValue": undefined,
                  "directives": Array [],
                  "kind": "VariableDefinition",
                  "type": Object {
                    "kind": "NonNullType",
                    "type": Object {
                      "kind": "NamedType",
                      "name": Object {
                        "kind": "Name",
                        "value": "String",
                      },
                    },
                  },
                  "variable": Object {
                    "kind": "Variable",
                    "name": Object {
                      "kind": "Name",
                      "value": "name",
                    },
                  },
                },
                Object {
                  "defaultValue": undefined,
                  "directives": Array [],
                  "kind": "VariableDefinition",
                  "type": Object {
                    "kind": "NamedType",
                    "name": Object {
                      "kind": "Name",
                      "value": "String",
                    },
                  },
                  "variable": Object {
                    "kind": "Variable",
                    "name": Object {
                      "kind": "Name",
                      "value": "url",
                    },
                  },
                },
                Object {
                  "defaultValue": undefined,
                  "directives": Array [],
                  "kind": "VariableDefinition",
                  "type": Object {
                    "kind": "NonNullType",
                    "type": Object {
                      "kind": "NamedType",
                      "name": Object {
                        "kind": "Name",
                        "value": "uuid",
                      },
                    },
                  },
                  "variable": Object {
                    "kind": "Variable",
                    "name": Object {
                      "kind": "Name",
                      "value": "id",
                    },
                  },
                },
              ],
            },
          ],
          "kind": "Document",
          "loc": Object {
            "end": 220,
            "start": 0,
          },
        }
      }
      refetchQueries={
        Array [
          Object {
            "query": Object {
              "definitions": Array [
                Object {
                  "directives": Array [],
                  "kind": "OperationDefinition",
                  "name": undefined,
                  "operation": "query",
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "repository",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "id",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "name",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "username",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "url",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "configurations",
                              },
                              "selectionSet": Object {
                                "kind": "SelectionSet",
                                "selections": Array [
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "id",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "name",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "configurationType",
                                    },
                                    "selectionSet": Object {
                                      "kind": "SelectionSet",
                                      "selections": Array [
                                        Object {
                                          "alias": undefined,
                                          "arguments": Array [],
                                          "directives": Array [],
                                          "kind": "Field",
                                          "name": Object {
                                            "kind": "Name",
                                            "value": "name",
                                          },
                                          "selectionSet": undefined,
                                        },
                                      ],
                                    },
                                  },
                                ],
                              },
                            },
                          ],
                        },
                      },
                    ],
                  },
                  "variableDefinitions": Array [],
                },
              ],
              "kind": "Document",
              "loc": Object {
                "end": 155,
                "start": 0,
              },
            },
          },
        ]
      }
    >
      <Component />
    </Mutation>
  </WithStyles(Drawer)>
</div>
`;
